source/logrepl/combined.go

```diff
@@ -0,0 +1,177 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logrepl
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/conduitio/conduit-connector-postgres/source/position"
+	"github.com/conduitio/conduit-connector-postgres/source/snapshot"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/jackc/pgx/v5/pgxpool"
+)
+
+type iterator interface {
+	Next(context.Context) (sdk.Record, error)
+	Ack(context.Context, sdk.Position) error
+	Teardown(context.Context) error
+}
+
+type CombinedIterator struct {
+	cdcIterator      *CDCIterator
+	snapshotIterator *snapshot.Iterator
+
+	activeIterator iterator
+}
+
+type Config struct {
+	Position        sdk.Position
+	SlotName        string
+	PublicationName string
+	Tables          []string
+	TableKeys       map[string]string
+	WithSnapshot    bool
+}
+
+func NewCombinedIterator(ctx context.Context, pool *pgxpool.Pool, conf Config) (*CombinedIterator, error) {
+	pos, err := position.ParseSDKPosition(conf.Position)
+	if err != nil {
+		sdk.Logger(ctx).Debug().
+			Err(err).
+			Msgf("failed to parse position: %s", string(conf.Position))
+
+		return nil, fmt.Errorf("failed to create logrepl iterator: %w", err)
+	}
+
+	if err := conf.Validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate longrepl config: %w", err)
+	}
+
+	lsn, err := pos.LSN()
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse LSN in position: %w", err)
+	}
+
+	cdcIterator, err := NewCDCIterator(ctx, &pool.Config().ConnConfig.Config, CDCConfig{
+		LSN:             lsn,
+		SlotName:        conf.SlotName,
+		PublicationName: conf.PublicationName,
+		Tables:          conf.Tables,
+		TableKeys:       conf.TableKeys,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CDC iterator: %w", err)
+	}
+
+	c := &CombinedIterator{
```

Comment:
IIUC, the logic should be: if `WithSnapshot` is true, and the position type is `Snapshot`, that means that the snapshot is required, and it's not done, so we need to use the snapshot iterator. In all other cases (snapshot not required, snapshot required but done), we activate the CDC iterator. Is that correct?

Reply:
`WithSnapshot` is the config value representing that this connector should take a snapshot, so the value won't change. Thus there are two cases:

* With Snapshot enabled
  * The position can be unknown or initial - use snapshot iterator
  * The position can be snapshot (to resume snapshot) - use snapshot iterator
  * The position can be CDC (snapshot completed and CDC has taken over) - use cdc iterator
* With snapshot disabled
  * It is only CDC mode


Does that make sense?

Reply:
I also had to check the logic a few times to understand it, the rate of [WTFs/minute](https://i0.wp.com/commadot.com/wp-content/uploads/2009/02/wtf.png) was a bit higher here, maybe it would be worth refactoring to make this flow a bit clearer.

Reply:
Yea, I will. It is somewhat annoying because in all cases the CDC iterator needs to be created. The snapshot iterator in this turn is optional in few cases.

Reply:
I agree with Lovro, it would be really good if the code would follow the cases you mentioned more closely. Plus, if possible, removing the `initial` type, since it's synonymous with position being `nil`.

Reply:
I refactored this a bit here, 7d40dad

It removes a bit of the mental load around initialization.
* The CDC iterator needs to always be initialized (and maybe started)
* The Snapshot iterator only needs to be initialized when snapshotting is enabled and CDC was not used last.

> Plus, if possible, removing the initial type, since it's synonymous with position being nil.

The initial position designates no state. Since the position cannot be nil, since it is a value type, we need to assign some meaning to the default value, this is why initial exists. Not sure how to communicate this in another way, besides using ptr or explicitly checking for the default value.